warp10/src/main/java/io/warp10/script/functions/ECRECOVER.java

```diff
@@ -0,0 +1,234 @@
+      // 0220 Integer on 0x20 bytes
+      //      7d97908b4472fd28d9381d795e9da95e07c5fd3eaea532d20a495ef543d51135
+      // 0220 Integer on 0x20 bytes
+      //      6db3861916bcd7096b9d701b44ef3cd2a74d087405ee36f01a857c1363ccb383
+      if (!(params.get(KEY_SIG) instanceof byte[])) {
+        throw new WarpScriptException(getName() + " invalid '" + KEY_SIG + "', expected BYTES.");
+      }
+
+      byte[] sig = (byte[]) params.get(KEY_SIG);
+
+      int offset = 0;
+
+      if (0 != (sig[1] & 0x80)) { // Length is encoded by sig[1] - 1 bytes
+        offset = 2;
+        int sbytes = (sig[1] & 0xFF) - 0x80;
+        int len = 0;
+        while(sbytes > 0) {
+          len = len << 8;
+          len |= sig[offset++] & 0xFF;
+          sbytes--;
+        }
+      } else {
+        offset = 2;
+      }
+
+      offset++; // Skip over '0x02'
+
+      // We assume that the size of r and s is less than 128 bytes (1024 bits)
+
+      int rlen = (int) sig[offset++];
+
+      r = new BigInteger(1, Arrays.copyOfRange(sig, offset, offset + rlen));
+      offset += rlen;
+      offset += 2;
+      s = new BigInteger(1, Arrays.copyOfRange(sig, offset, sig.length));
+      if (r.compareTo(spec.getN()) > 0 || r.compareTo(BigInteger.ONE) < 0) {
+        throw new WarpScriptException(getName() + " invalid value for r, should be in [1, 0x00" + spec.getN().toString(16) + "] but was 0x" + r.toString(16) + ".");
+      }
+      if (s.compareTo(spec.getN()) > 0 || s.compareTo(BigInteger.ONE) < 0) {
+        throw new WarpScriptException(getName() + " invalid value for s, should be in [1, 0x00" + spec.getN().toString(16) + "] but was 0x" + s.toString(16) + ".");
+      }
+    } else if (null != params.get(KEY_R) && null != params.get(KEY_S)) {
+      if (!(params.get(KEY_R) instanceof String)) {
+        throw new WarpScriptException(getName() + " invalid '" + KEY_R + "', expected STRING.");
+      }
+
+      if (!(params.get(KEY_S) instanceof String)) {
+        throw new WarpScriptException(getName() + " invalid '" + KEY_S + "', expected STRING.");
+      }
+
+      String str = ((String) params.get(KEY_R)).toLowerCase();
+
+      if (str.startsWith("0x")) {
+        r = new BigInteger("00" + str.substring(2), 16);
+      } else {
+        r = new BigInteger(str);
+      }
+
+      str = ((String) params.get(KEY_S)).toLowerCase();
+
+      if (str.startsWith("0x")) {
+        s = new BigInteger("00" + str.substring(2), 16);
+      } else {
+        s = new BigInteger(str);
+      }
+    } else {
+      throw new WarpScriptException(getName() + " expects '" + KEY_SIG + "' or '" + KEY_R + "' and '" + KEY_S + "' to be provided.");
+    }
+
+    int nbits = spec.getN().bitLength();
+    byte[] hash = (byte[]) params.get(KEY_HASH);
+
+    z = new BigInteger(1, hash);
+
+    if (nbits < hash.length * 8) {
+      z = z.shiftRight(hash.length * 8 - nbits);
+    }
+
+    List<Object> candidates = new ArrayList<Object>();
+
+    for (int j = 0; j < H.intValue(); j++) {
```

Comment:
It can take more than 3 minutes to execute this loop (tested with c2pnb368w1 and SHA1). We should protect this function with either capabilities or authentication to avoid abuse.